tiledb/type/range/range.h

```diff
@@ -257,6 +259,85 @@ class Range {
   uint64_t partition_depth_;
 };
 
+/**
+ * Performs checks to verify a range is a subset of the superset.
+ *
+ * Both the superset and the range must be valid ranges of the appropriate type
+ * as checked by ``check_range_is_valid``.
+ *
+ * @param superset The range that is the assumed superset. Must be a valid range
+ * with data of type T.
+ * @param range The range that is the assumed subset. Must be a valid
+ * range with data of type T.
+ * @return Status of the checks.
+ */
+template <
+    typename T,
+    typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+Status check_range_is_subset(const Range& superset, const Range& range) {
+  auto domain = (const T*)superset.data();
+  auto r = (const T*)range.data();
+  if (r[0] < domain[0] || r[1] > domain[1]) {
+    std::stringstream ss;
+    ss << "Range [" << r[0] << ", " << r[1] << "] is out of domain bounds ["
+       << domain[0] << ", " << domain[1] << "]";
+    return Status_RangeError(ss.str());
+  }
+  return Status::Ok();
+};
+
+/**
+ * Performs correctness checks for a valid range. If any validity checks fail
+ * an error status is returned.
+ *
+ * @param range The range to check.
+ * @return Status of the checks.
+ */
+template <
+    typename T,
+    typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+Status check_range_is_valid(const Range& range) {
+  // Check has data.
+  if (range.empty())
+    return Status_RangeError("Range is empty");
+  auto r = (const T*)range.data();
+  // Check for NaN
+  if constexpr (std::is_floating_point_v<T>) {
+    if (std::isnan(r[0]) || std::isnan(r[1]))
+      return Status_RangeError("Range contains NaN");
+  }
+  // Check range bounds
+  if (r[0] > r[1]) {
+    std::stringstream ss;
+    ss << "Lower range bound " << r[0]
+       << " cannot be larger than the higher bound " << r[1];
+    return Status_RangeError(ss.str());
+  }
+  // Return okay
+  return Status::Ok();
+};
+
+/**
+ * Crop a range to the requested bounds
+ *
+ * @param bounds Bounds to crop the range to. Must be a valid range with data of
+ * type T.
+ * @param range The range to crop. Must be a valid range with data of type T.
+ * @return Status that returns an error if range was mutated.
+ */
+template <
+    typename T,
+    typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+void crop_range(const Range& bounds, Range& range) {
```

Comment:
I'm assuming that the follow-on PR you anticipated will use this to replace the `Dimension::crop_range_func`? But, taking a step back: do you have a use-case for this crop function?

Currently, it seems to only be called through one code path in dimension.cc (via `Domain::crop_ndrange`). If you don't have a use-case, I am inclined to isolate it (and ideally refactor to combine with the very similar `adjust_range_oob` function there) than keeping it here in Range. Any reduction in scope of Range is a win IMO.

Reply:
This function will replace the `adjust_range_oob` function that is defined in `Dimension` and only used when adding ranges to `Subarray`. I can move it directly the the `RangeSetAndSuperset` class or I can leave it here. I am planning on removing `adjust_range_oob` from dimension.

Reply:
From our conversation: this will be used for `Domain::crop_ndrange`, `Subarray::add_range`, and `Subarray::add_label_range`. We are leaving it in `range.h` since it is needed in multiple code paths.